Base API for chaos-testing

dodokek opened the pull request:
Implemented base functions for K6 plugin such as deploying chaos-experiment and checking cluster health.

dodokek committed:
Implemented base functionality for chaos testing, add simple k8s API tests

Fix naming

potemin1999 commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+export function teardown(opts) {
+  // Creating another instance of controller to
```
I suggest to create a single `ChaosController` from within `setup` function (as all of them use the same namespace).  
The `teardown` function just calls `clearExperimentData` on already existing controller.  
This also will fix a symmetry where `setup` creates resources, and `teardown` destroys them (instead of creating one in `killSlave`)

potemin1999 commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+// The killPod function terminates a pod within a Kubernetes cluster according to specifications provided.
+export function killSlave(namespace) {
```
This function is used only once, how about to move its code here? After `setup`+`teardown` is balanced, it is just 5 SLOCs

potemin1999 commented on examples/master-network-loss.js:
```diff
@@ -0,0 +1,67 @@
+  if (chaosController == null) {
+    fail("couldn't create chaos controller instance")
+  }
+  console.log("Stargin exp")
```
```suggestion
  console.log("Starting experiment")
```

potemin1999 commented on examples/utils.js:
```diff
@@ -0,0 +1,68 @@
+}
+
+export function getBaseUrl() {
+  let node = typeof __ITER == "undefined" ? vars.hosts[0] : vars.hosts[__ITER % vars.hosts.length];
```
Nit: this makes things clearer (that base url is round-robined when running test iterations or 0 as fallback option)
```suggestion
  const index = typeof __ITER === "undefined" ? 0 : __ITER;
  const node = vars.hosts[index % vars.hosts.length];
```

potemin1999 commented on examples/utils.js:
```diff
@@ -0,0 +1,68 @@
+  return `http://${node}`;
+}
+
+function getAuthToken(expired_at) {
```
Function arguments are `camelCase`. And it is unused btw.
```suggestion
function getAuthToken(expiresAt) {
```
Simple [Google's TS naming conventions](https://google.github.io/styleguide/tsguide.html#naming-rules-by-identifier-type) may be helpful

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+)
+
+type ChaosController struct {
+	cluster_config_path string
```
```suggestion
	clusterConfigPath string
```
[Effective Go > Names > Mixed Caps](https://go.dev/doc/effective_go#names) reference.

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+		return nil
+	}
+
+	config_path := getConfigPath()
```
```suggestion
	configPath := getConfigPath()
```

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+		return err.Error()
+	}
+
+	err = cc.CheckPodsHealth(allPods)
```
nit:
```suggestion
	if err = cc.CheckPodsHealth(allPods); err != nil { 
```

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+func (cc *ChaosController) CheckPodsHealth(pods []string) error {
+	cur_namespace := extractNamespace(cc.curExp)
```
nit:
```suggestion
	curNamespace := extractNamespace(cc.curExp)
```

potemin1999 commented on pkg/k8s/config/config.go:
```diff
@@ -0,0 +1,26 @@
+// GetConfig creates a new k8s config
+func GetConfig() *rest.Config {
+	configPath := GetConfigPath()
+	config, _ := clientcmd.BuildConfigFromFlags("", configPath)
```
Does the config at "~/.kube/config" always exist?    
If no, then there is a potential error unhandled - building config from file which could not be read.

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+
+// Get all podes, affected by chaos experiment and reload them
+// This is done in order to avoid CrashLoopBackOff pod status
+func RestartAffectedPods(cc *ChaosController) error {
```
Move the receiver before function name
```suggestion
func (cc *ChaosController) RestartAffectedPods() error {
```

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+// Wait cluster to reconfigure after chaos experiment
+func (cc *ChaosController) ExperimentSleep(durationStr string) error {
```
Does not need to be a method
```suggestion
func ExperimentSleep(durationStr string) error {
```

potemin1999 commented on internal/k8s/pods/pods.go:
```diff
@@ -0,0 +1,51 @@
+func (pods *Pods) KillByName(ctx context.Context, namespace string, podName string) error {
+	podsInNamespace := pods.Client.CoreV1().Pods(namespace)
+	err := podsInNamespace.Delete(ctx, podName, v1.DeleteOptions{})
+	return err
```
nit:
```suggestion
	return pods.Client.CoreV1().Pods(namespace).Delete(ctx, podName, v1.DeleteOptions{})
```

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+	"k8s.io/client-go/util/homedir"
+)
+
+type ChaosController struct {
```
nit: structure changes:
1. Rename package `chaos_controller` to `chaos`
2. Extract functions from `chaos/chaosmesh.go` to `chaos/utils.go`
3. Rename `chaosmesh.go` to `controller.go` (file will be about controller)

potemin1999 commented on examples/utils.js:
```diff
@@ -0,0 +1,68 @@
+  let resp = http.put(`${getBaseUrl()}/admin/bauth/users/test:perf/password`, '{ "password": "pwd" }');
+  respCheck(resp, { httpCode: 201, bastidaCode: null });
+
+  resp = http.put(`${getBaseUrl()}/admin/bauth/users/test:perf/token`, '{ "expires_at": 8015414400 }');
```
Fixes unused:
```suggestion
  resp = http.put(`${getBaseUrl()}/admin/bauth/users/test:perf/token`, '{ "expires_at": ${expiresAt} }');
```

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+		context.TODO(),
+		v1.DeleteOptions{},
+		v1.ListOptions{},
+	)
```
Field `ChaosController.gvr` is write-only for now, and it already stores resource name.  

Either remove `cc.gvr = gvr` from `RunChaosExperiment`
Or reuse stored gvr (and run this function before starting next experiment):
```suggestion
func (cc *ChaosController) ClearLastExperimentData() {
        if cc.gvr == nil { return }
	err := cc.dynamicClient.Resource(cc.gvr).Namespace(cc.namespace).DeleteCollection(
		context.TODO(),
		v1.DeleteOptions{},
		v1.ListOptions{},
	)
        cc.gvr = nil
```

potemin1999 commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+// recoveryTime - how much time to wait for cluster to recover after experiment
+func (cc *ChaosController) RunChaosExperiment(configPath string, restartNodes bool, recoveryTime string) error {
+	// Work with experiment yaml file
+	experiment_file, err := os.ReadFile(configPath)
```
nit:
```suggestion
	experimentFile, err := os.ReadFile(configPath)
```

potemin1999 commented on internal/experiments/experiments.go:
```diff
@@ -0,0 +1,58 @@
+	"xk6-khorne/internal/experiments/chaos_controller"
+
+	"github.com/dop251/goja"
+	"go.k6.io/k6/js/modules"
```
For throwing errors:
```suggestion
	"go.k6.io/k6/js/modules"
	"go.k6.io/k6/js/common"
	"errors"
```

potemin1999 commented on internal/experiments/experiments.go:
```diff
@@ -0,0 +1,58 @@
+
+	// parse cmd. line arguments
+	args := call.Arguments
+	var namespace string
+	if len(args) > 0 {
+		namespace = args[0].String()
+	} else {
+		log.Fatalf("no arguments passed to chaos controller constructor")
+		return nil
+	}
```
Throwing error to JS code:
```suggestion
        if len(args) != 1 {
            common.Throw(rt, errors.New("ChaosController constructor expects 1 argument")
        }
        namespace := args[0].String()
```

lowitea commented on chaos.go:
```diff
@@ -0,0 +1,24 @@
+package chaos
+
+import (
+	"fmt"
+
+	_ "xk6-khorne/internal/experiments" // Register the experiments module as well
+	_ "xk6-khorne/internal/k8s"         // Register the k8s module as well
+
+	"go.k6.io/k6/js/modules"
+)
+
+const version = "v0.0.2"
+
+func init() {
+	modules.Register("k6/x/chaos", &Chaos{
+		Version: version,
+	})
+	fmt.Println("Running k6io/xk6-chaos@$" + version)
+}
+
+// Chaos is the main export of the chaos engineering extension
+type Chaos struct {
+	Version string
+}
```
Why do you call the module Chaos here, even though the plugin is called Khorne?

lowitea commented on go.mod:
```diff
@@ -0,0 +1,98 @@
+module xk6-khorne
```
As far as I know, the module should have the prefix "github.com/picodata/"

lowitea commented on go.mod:
```diff
@@ -0,0 +1,98 @@
+	k8s.io/api v0.31.1
+	k8s.io/apimachinery v0.31.1
+	k8s.io/client-go v0.31.1
+)
```
Please run `go mod tidy`

lowitea commented on chaos.go:
```diff
@@ -0,0 +1,24 @@
+	"go.k6.io/k6/js/modules"
+)
+
+const version = "v0.0.2"
```
In such projects, it is better to follow semantic versioning https://semver.org/

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+
+
+import { ChaosController } from 'k6/x/chaos/experiments';
+import { textSummary } from 'https://jslib.k6.io/k6-summary/0.0.1/index.js';
```
Not used

lowitea commented on examples/master-network-loss.js:
```diff
@@ -0,0 +1,67 @@
+
+
+import { ChaosController } from 'k6/x/chaos/experiments';
+import { textSummary } from 'https://jslib.k6.io/k6-summary/0.0.1/index.js';
```
Not used

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+  }
+}
+
+
```
I recommend naming the file `kill_one_slave_node.js`. And similarly related files.

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+    fail("couldn't create chaos controller instance")
+  }
+
+  chaosController.runChaosExperiment("./examples/chaosmesh/1_slave_kill.yaml", NO_RECOVER, "40s");
```
Because of the naming, it is unclear whether you enable recovery here or disable it

lowitea commented on examples/utils.js:
```diff
@@ -0,0 +1,68 @@
+    "Content-Type": "application/json",
+    "Bauth-token": getAuthToken(expired_at),
+  };
+}
```
You can remove it from the example. These are special headers for the project.

lowitea commented on examples/utils.js:
```diff
@@ -0,0 +1,68 @@
+  console.info(`Generated random seed: ${seed}`);
+}
+
+export function respCheck(resp, opts) {
```
The check function was created for one specific project and you don't need it here

lowitea commented on examples/utils.js:
```diff
@@ -0,0 +1,68 @@
+import { check, fail } from "k6";
```
I think you can drop the entire file.

lowitea commented on examples/vars.js:
```diff
@@ -0,0 +1,11 @@
+export default {
```
For the examples, this file is redundant.

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+
+export const options = Object.assign({}, vars.generalOpts, {
+  vus: 1,
+  duration: "1s",
```
Maybe you need `iterations` option?

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+  }
+  chaosController.clearExperimentData("podchaos")
+
+  respCheck(http.del(`${getBaseUrl()}/ucp/0001/perf/plain`, null, opts.httpParams));
```
Use more simple and general path, maybe `/check/path`?

lowitea replied:
Although the deletion seems superfluous here.

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+
+export default function (opts) {
+  let uid = exec.scenario.iterationInTest * (opts.startId || 1);
+  let data = `{"id":0,"params":[{"uid":${uid},"p1":1,"p2":2}]}`;
```
Please use simpler and more general example.

lowitea commented on examples/1_slave_kill.js:
```diff
@@ -0,0 +1,70 @@
+
+  killSlave("test-namespace")
+
+  let resp = http.put(`${getBaseUrl()}/ucp/0001/perf/plain/uid/${uid}`, data, opts.httpParams);
```
Please use simpler uri

lowitea commented on examples/master-network-loss.js:
```diff
@@ -0,0 +1,67 @@
+// This test checks cluster tolerance to pod failures
```
In this file, please take into account all the comments from `1_slave_kill.js` file.

lowitea replied:
It is better to use dashes and underscores in file names uniformly.

lowitea commented on examples/master-network-loss.js:
```diff
@@ -0,0 +1,67 @@
+// This test checks cluster tolerance to pod failures
+// We basically disable one slave node from storage replicaset
```
master node?

lowitea commented on internal/k8s/k8s.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+func init() {
+	modules.Register("k6/x/chaos/k8s", K8sRoot{})
```
```suggestion
	modules.Register("k6/x/khorne/k8s", K8sRoot{})
```

lowitea commented on internal/experiments/experiments.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+func init() {
+	modules.Register("k6/x/chaos/experiments", &ExperimentsRoot{})
```
```suggestion
	modules.Register("k6/x/khorne/experiments", &ExperimentsRoot{})
```

lowitea commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+	p := pods.New(c)
+	if err != nil {
+		return nil
+	}
```
It is usually better to check the error immediately after the place where it was received.

lowitea commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+	}
+	reader := bytes.NewReader(experiment_file)
+	decoder := yaml.NewYAMLOrJSONDecoder(reader, 1000)
+	parsedYaml := &unstructured.Unstructured{}
```
Maybe you can use this struct https://github.com/chaos-mesh/chaos-mesh/blob/release-2.7/api/v1alpha1/networkchaos_types.go#L31 for parsing?

dodokek replied:
I think so, however existing structure seems more general. There will be another chaos types (podChaos for example).

lowitea commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == coreV1.PodConditionType(coreV1.NodeReady) {
+			if condition.Status != coreV1.ConditionTrue {
+				return fmt.Errorf("node %s is not ready. Status: %s, Reason: %s, Message: %s",
+					podName, condition.Status, condition.Reason, condition.Message)
+			}
```
Highlighting the beginning and end of a block greatly simplifies the reading of the code structure.

```suggestion
			if condition.Status != coreV1.ConditionTrue {
				return fmt.Errorf("node %s is not ready. Status: %s, Reason: %s, Message: %s",
					podName, condition.Status, condition.Reason, condition.Message
				)
			}
```

lowitea commented on internal/experiments/chaos_controller/chaosmesh.go:
```diff
@@ -0,0 +1,232 @@
+
+// getConfigPath fetches the path to the users kubeconfig
+func getConfigPath() string {
+	if configPath := os.Getenv("K6_CHAOS_KUBECONFIG"); configPath != "" {
```
```suggestion
	if configPath := os.Getenv("KHORNE_KUBECONFIG"); configPath != "" {
```

dodokek committed:
Adjust test for k8s client

dodokek committed:
Change folder structure, refactor code

Get rid of complex folder structure. Make a bunch of callable independent functions instead of methods of one structure

dodokek committed:
Improve error handling

dodokek committed:
Add K6 example for splitbrain

dodokek committed:
Remove dead code, cosmetic changes

dodokek committed:
Update README

lowitea commented on examples/kill_one_slave_node.js:
```diff
@@ -0,0 +1,26 @@
+
+export const options = Object.assign({}, {}, {
+  vus: 1,
+  duration: "50s",
```
It seems that you don't need the duration setting here, the test will end as soon as it exits the default function

lowitea commented on examples/splitbrain_simulation.js:
```diff
@@ -0,0 +1,29 @@
+// This test checks cluster tolerance to pod failures
+// We basically disable one slave node from storage replicaset
+// And hope that replicaset will return in consistent state
```
Is this exactly the right description?

lowitea commented on examples/splitbrain_simulation.js:
```diff
@@ -0,0 +1,29 @@
+
+export const options = Object.assign({}, {}, {
+  vus: 1,
+  duration: "50s",
```
Unnecessary duration

lowitea commented on examples/kill_one_slave_node.js:
```diff
@@ -0,0 +1,26 @@
+import k6 from 'k6';
+import khorne from "k6/x/khorne";
+
+export const options = Object.assign({}, {}, {
```
You don't need to use assign function here

lowitea commented on examples/splitbrain_simulation.js:
```diff
@@ -0,0 +1,29 @@
+import k6 from 'k6';
+import khorne from "k6/x/khorne";
+
+export const options = Object.assign({}, {}, {
```
You don't need to use assign function here

lowitea commented on examples/chaosmesh/kill_master_node.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: chaos-mesh.org/v1alpha1
+metadata:
+  namespace: test-namespace
+  name: slave-failure
```
master?

lowitea commented on examples/splitbrain_simulation.js:
```diff
@@ -0,0 +1,29 @@
+  iterations: 1
+});
+
+export default function (opts) {
+  khorne.RunChaosExperiment("test-namespace", "./examples/chaosmesh/kill_master_node.yaml")
+  khorne.ExperimentSleep("20s")
+  khorne.ClearChaosCache("test-namespace")
+  let result = khorne.CheckPodsHealth("test-namespace", ["storage-0-1"])
+
+  if (result.success) {
+    k6.fail("Master node became consistent, however new master has been elected. Splitbrain wasn't detected. " + result.error)
+  } else {
+    console.log("Splitbrain has been handled, old master status: " + result.error)
+  }
+
+}
```
It seems to me that this test does not check the real splitbrain. It looks like it checks the node recovery after restart and failover.

It's OK. But you need to correct the description and name of the test.

lowitea commented on examples/splitbrain_simulation.js:
```diff
@@ -0,0 +1,29 @@
+  }
+
+}
+
+
+
```
Please configure prettier for your IDE and use it to format code.

lowitea commented on src/chaos/utils.go:
```diff
@@ -0,0 +1,17 @@
+	"time"
+)
+
+// Wait cluster to reconfigure after chaos experiment
+func ExperimentSleep(durationStr string) error {
+	experimentDuration, err := time.ParseDuration(durationStr)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(experimentDuration)
+
+	return nil
+}
```
Maybe we should just name this function Sleep?

dodokek replied:
Yep

lowitea commented on src/client/client.go:
```diff
@@ -0,0 +1,39 @@
+	"k8s.io/client-go/util/homedir"
+)
+
+// GetConfigPath fetches the path to the users kubeconfig
+func GetConfigPath() string {
```
I think this function should be private

lowitea commented on src/client/client.go:
```diff
@@ -0,0 +1,39 @@
+package client
```
Why is this package named "client" if its only function is to get kubectl config?

lowitea commented on src/chaos/k8s_interaction_test.go:
```diff
@@ -0,0 +1,97 @@
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+func TestCheckUnhealthyNodeValidation(t *testing.T) {
```
I tried to run go test, but I got an error message

```
--- FAIL: TestCheckUnhealthyNodeValidation (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xa0c76a]
```

dodokek replied:
Somehow it works on my Ubuntu 22.04, go version go1.22.0 linux/amd64. Will need a closer look

dodokek replied:
Figured it out. This happened because there was no ~/.kube/config file on your machine. I've changed logic so there is no error now

lowitea commented on src/client/client.go:
```diff
@@ -0,0 +1,39 @@
+	"os"
+	"path/filepath"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
```
Do you really need this?

lowitea commented on src/chaos/k8s_interaction_test.go:
```diff
@@ -0,0 +1,97 @@
+	}
+}
+
+func TestCheckHealthyNodeValidation(t *testing.T) {
```
And this test failed too

lowitea commented on src/chaos/k8s_interaction_test.go:
```diff
@@ -0,0 +1,97 @@
+
+func TestCheckUnhealthyNodeValidation(t *testing.T) {
+	clientset := fake.NewSimpleDynamicClient(scheme.Scheme)
+	// Create a fake dynamic client
```
This comment should be above

lowitea commented on src/chaos/k8s_interaction_test.go:
```diff
@@ -0,0 +1,97 @@
+
+func TestCheckHealthyNodeValidation(t *testing.T) {
+	clientset := fake.NewSimpleDynamicClient(scheme.Scheme)
+	// Create a fake dynamic client
```
This comment should be above

lowitea commented on src/chaos/controller.go:
```diff
@@ -0,0 +1,210 @@
+	Namespace     string
+}
+
+// NewController creates a new instance of ChaosController
+func NewController(namespace string) *ChaosController {
+	config, err := client.GetConfig()
+	if err != nil {
+		log.Fatal(err.Error())
+		return nil
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Failed to create Kubernetes client: %v", err)
+	}
+
+	return &ChaosController{dynamicClient, namespace}
+}
```
nit: I don't think this will significantly affect performance, but I think it's still better to use a singleton here.

```suggestion
var chaosControllerSingltone *ChaosController

// NewController creates a new instance of ChaosController
func NewController(namespace string) *ChaosController {
	if chaosControllerSingltone != nil {
		return chaosControllerSingltone
	}

	config, err := client.GetConfig()
	if err != nil {
		log.Fatal(err.Error())
		return nil
	}

	dynamicClient, err := dynamic.NewForConfig(config)
	if err != nil {
		log.Fatalf("Failed to create Kubernetes client: %v", err)
	}

	chaosControllerSingltone = &ChaosController{dynamicClient, namespace}

	return chaosControllerSingltone
}
```

lowitea commented on src/chaos/controller.go:
```diff
@@ -0,0 +1,210 @@
+
+// Deleting the experiment from chaosmesh list
+// to avoid collisions, when the same experiment is started
+func ClearChaosCache(namespace string) error {
```
Maybe it's better to choose the name `StopAllChaosExperiments`?

dodokek replied:
Chaos experiment stops by itself when duration time has passed. After experiment finished, in k8s scope lays some metadata about passed experiment. This metadata won't allow to run experiment with name ```master-failure``` twice in a row, for example. Thus, we need to clear this sort of Cache or Metadata (don't know which name is better) to run the same chaos test twice.

So I think better name would be ClearChaosCache or ClearChaosmechMeta or smth like that

lowitea commented on src/chaos/controller.go:
```diff
@@ -0,0 +1,210 @@
+		}
+	}
+
+	return nil
```
Maybe it's more logical to return `true, nil` here?

dodokek replied:
The result, visible to user will be, exactly as you pointed, a structure {true, nil}. Return statement, visible for users are defined in file khorne.go
```go
func (e *Chaos) Exports() modules.Exports {
	return modules.Exports{
		Named: map[string]interface{}{
			"ExperimentSleep":    e.ExperimentSleep,
			"RunChaosExperiment": e.RunChaosExperiment,
			"ClearChaosCache":    e.ClearChaosCache,
			"CheckClusterHealth": e.CheckClusterHealth,
			"CheckPodsHealth":    e.CheckPodsHealth,
			"RestartPods":        e.RestartPods,
		},
	}
}

func (p *Chaos) CheckClusterHealth(namespace string) Result {
	if err := chaos.CheckClusterHealth(namespace); err != nil {
		return Result{false, err.Error()}
	}
	return Result{Success: true}
}
```

dodokek committed:
Update formatting and comments

Additionaly made minor changes in contoller code, adding singletone for chaos controller instance

dodokek committed:
Fix k8s tests

dodokek committed:
Add k8s API tests to github actions

lowitea commented on examples/kill_master_node.js:
```diff
@@ -0,0 +1,35 @@
+import k6 from "k6";
+import khorne from "k6/x/khorne";
+
+export const options = Object(
+  {},
+  {},
+  {
+    vus: 1,
+    iterations: 1,
+  }
+);
```
You can do it just like that:

```suggestion
export const options = {
  vus: 1,
  iterations: 1,
};
```

lowitea commented on examples/kill_one_slave_node.js:
```diff
@@ -0,0 +1,29 @@
+import k6 from "k6";
+import khorne from "k6/x/khorne";
+
+export const options = Object(
+  {},
+  {},
+  {
+    vus: 1,
+    iterations: 1,
+  }
+);
```
And this

```suggestion
export const options = {
  vus: 1,
  iterations: 1,
};
```

dodokek committed:
Remove unused config creation

dodokek committed:
Simplify JS var declatation
